tmedia 689 autoplay carousel

JackHowa opened the pull request:
## Ticket

- [TMEDIA-689](https://arcpublishing.atlassian.net/browse/TMEDIA-689) sub-task of https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TMEDIA-689&assignee=5e387d6a1b1d910e5dfd7a9b

## Description

Add autoplay and custom button functionality

## Acceptance Criteria

- Autoplay functionality opt-in and -out on arrival to the page of the carousel
- Feature is accessible by detecting reduced motion 

## Test Steps

1. Checkout branch - `git checkout TMEDIA-689-autoplay-gallery`
2. Run Storybook `npm run storybook`
3. Look at the storybook for autoplay functionality 
![Screen Shot 2022-04-18 at 17 00 03](https://user-images.githubusercontent.com/5950956/163884389-9a02bef9-ea62-4b94-981b-a2302634a26a.png)
4. Notice "Start Autoplay" on first arrival to a story with `enableAutoplay`
5. Click `Start Autoplay`
6. Wait 4 seconds (default from engine theme sdk). Then you will go to the next slide. You will also see the "Stop Autoplay" available 
![Screen Shot 2022-04-18 at 17 01 32](https://user-images.githubusercontent.com/5950956/163884574-1cc5da8e-1414-4eae-a0ac-c52e70a79a17.png)
8. Go to the custom button story example to see the buttons with our own custom styling work as before 
![Screen Shot 2022-04-18 at 17 02 44](https://user-images.githubusercontent.com/5950956/163884738-e1318982-f16e-4ae1-9c25-c48160b7c8a9.png)

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**


-----

## Notes 

- Checking for reduced motion per accessibility requirement https://www.w3.org/TR/wai-aria-practices-1.2/examples/carousel/carousel-1-prev-next.html

## Questions 


- The engine theme sdk does not let you set the interval duration. I didn't implement this feature but set the 4000 from engine theme sdk interval along with its same logic. Is this ok to leave out? 
- Should we take in autoplay phrases as well as custom buttons? I think that the buttons should be enough customization. Thoughts? This follows the same pattern as the next and previous and fullscreen buttons. 

## Todos 

- [x] Take in autoplay components in storybook
- [x] Include autoplay custom buttons in proptypes 
- [x] Describe feature in carousel docs
- [x] Add tests. Include mocked reduced motion 
- [x] Test clicking to initiate the autoplay

JackHowa committed:
feat: Add autoplay functionality with accessibility check for carousel

JackHowa committed:
test: Ignore helper component via dan abramov

JackHowa committed:
feat: Set default autoplay interval

JackHowa committed:
docs: Fix typo with id

JackHowa committed:
style: Remove unnecessary ignores and comments

JackHowa committed:
feat: Remove reset at end of next slide

JackHowa committed:
test: Add reduce media mock and show the tests on click of the start/stop autoplay

JackHowa committed:
docs: Add proptypes and description for node of custom buttons

JackHowa committed:
docs: Add description with accessibility explanation for autoplay

JackHowa committed:
fix: Ensure typo does not affect render logic for resolve play button

JackHowa committed:
docs: Show custom buttons in storybook play/pause autoplay

JackHowa committed:
refactor: Standardize autoplay case

matthewroach commented on src/components/carousel/index.jsx:
```diff
@@ -115,6 +142,12 @@ const Carousel = ({
 
+	// a prefers-reduced-motion user setting must always override Autoplay
+	const AutoplayEnabledAndAllowed =
+		enableAutoplay && !!window.matchMedia("'(prefers-reduced-motion: reduce)");
```
`window` will throw an error on Fusion server side render and cause this code to become client side only. We should add an `isServerSide()` util function to handle checking for `window` being defined

JackHowa replied:
@matthewroach updated. thanks for the note

matthewroach commented on src/components/carousel/index.jsx:
```diff
@@ -182,6 +217,28 @@ const Carousel = ({
+
 	/* istanbul ignore next  */
 	const fullScreenEnabledAllowed =
 		(document.fullscreenEnabled || document.webkitFullscreenEnabled) && enableFullScreen;
```
`document` is not available on Fusion server side render and will throw an error as per the `window` reference above

JackHowa replied:
@matthewroach updated. thanks for the note

matthewroach reviewed (dismissed):
We also need to add `aria-live="off"` to the track `div` when auto play is running, and set it to `aria-live="polite"` when not in auto play mode

JackHowa committed:
fix: Set aria live based on autoplay status

Per https://github.com/WPMedia/arc-themes-components/pull/73#pullrequestreview-945309638

JackHowa committed:
fix: Add serverside check to ensure client-side availability of window and document

Per https://github.com/WPMedia/arc-themes-components/pull/73#discussion_r852766200 https://github.com/WPMedia/arc-themes-components/pull/73#discussion_r852767162

JackHowa committed:
fix: Ensure casing correct for serverside check

JackHowa commented:
@matthewroach updated based on feedback

JackHowa committed:
style: Fix naming typo

JackHowa committed:
feat: Export utils

JackHowa committed:
docs: Update prop description based on pr feedback to make it more clear that the autoplay enable is to show the button

JackHowa committed:
docs: Update storybook story names for clarity

matthewroach commented on src/components/carousel/index.jsx:
```diff
@@ -203,6 +264,8 @@ const Carousel = ({
 						{pageCountPhrase(slide, totalSlides) || `${slide} of ${totalSlides}`}
 					</p>
 				) : null}
+				{AutoplayEnabledAndAllowed && !isAutoplaying ? resolvedStartAutoplayButton : null}
```
Using two buttons - one for each interaction will require us to manage state and focus for the user - could we get away with using one and have the user set the label and icon?

JackHowa replied:
I was following the same custom button pattern as the next and previous buttons. What if the user wants to put an icon on the left or right side of the label? or do something else custom? should I add an icon left or right option as well? 

Ideally, we'd want a custom block developer to use our button component kind of like how I've done in the storybook

new props: 
- autoplayStartLabelText
- autoplayStartIcon
- autoplayStartIconLocation 
- autoplayStopLabelText
- autoplayStopIcon
- autoplayStopIconLocation

JackHowa replied:
I'm not totally sure how I could avoid using state here. what would determine whether to show the autoplayStartLabelText vs autoplayStopLabelText ? is this an instance of using aria attributes? @matthewroach 

looking to https://www.w3.org/TR/wai-aria-practices-1.2/examples/carousel/carousel-1-prev-next.html for inspiration

matthewroach replied:
I don't think we need so many options - what if we provide the ability for them to set the label and icon for each state for now?

nschubach replied:
Are you thinking of making the toggle button a callback where we pass in the current state and letting them decide?

JackHowa replied:
yep can do @matthewroach 

> I don't think we need so many options - what if we provide the ability for them to set the label and icon for each state for now?

JackHowa replied:
this has been updated @matthewroach

brentmiller1973 commented on src/components/carousel/index.stories.mdx:
```diff
@@ -362,3 +364,66 @@ const Feature = () => (
+			</Carousel.Item>
+		</Carousel>
+	</Story>
+</Canvas>
```
@JackHowa Its important, especially considering the complexity of this component to see the component with with multiple states enabled.  This will really help QA as well.  For autoplay, I want to see examples when their are multiple slides visible and when it's in fullscreen mode.

```
** Autoplay enabled with full screen**

<Canvas>
    <Story name="Autoplay enabled with full screen">
        <Carousel id="carousel-1" label="Carousel of Images" slidesToShow={1} enableAutoplay enableFullScreen>
            <Carousel.Item label="Slide 1 of 4">
                <a href="/">
                    <Image src="/computer-user.jpeg" alt="A person sat down using a laptop computer" />
                </a>
            </Carousel.Item>
            <Carousel.Item label="Slide 2 of 4">
                <Image src="/camera.jpeg" alt="A camera with photos in front of it" />
            </Carousel.Item>
            <Carousel.Item label="Slide 3 of 4">
                <Image src="/canyon.jpeg" alt="Landscape view of a canyon" />
            </Carousel.Item>
            <Carousel.Item label="Slide 4 of 4">
                <Image src="/coffee.jpeg" alt="A cup of coffee" />
            </Carousel.Item>
        </Carousel>
    </Story>
</Canvas>

** Autoplay enabled with multiple slides **

<Canvas>
    <Story name="Autoplay enabled with multiple slides">
        <Carousel
            id="carousel-1"
            label="Carousel of Images"
            enableAutoplay
            nextButton={
                <Carousel.Button id="carousel-1" label="Next Slide">
                    <Icon name="ArrowRight" />
                </Carousel.Button>
            }
            previousButton={
                <Carousel.Button id="carousel-1" label="Previous Slide">
                    <Icon name="ArrowLeft" />
                </Carousel.Button>
            }
        >
            <Carousel.Item label="Slide 1 of 6">
                {({ viewable }) => (
                    <a href="/" aria-hidden={viewable ? null : true} tabIndex={viewable ? null : "-1"}>
                        <Image src="/computer-user.jpeg" alt="A person sat down using a laptop computer" />
                    </a>
                )}
            </Carousel.Item>
            <Carousel.Item label="Slide 2 of 6">
                <Image src="/camera.jpeg" alt="A camera with photos in front of it" />
            </Carousel.Item>
            <Carousel.Item label="Slide 3 of 6">
                <Image src="/canyon.jpeg" alt="Landscape view of a canyon" />
            </Carousel.Item>
            <Carousel.Item label="Slide 4 of 6">
                <Image src="/coffee.jpeg" alt="A cup of coffee" />
            </Carousel.Item>
            <Carousel.Item label="Slide 5 of 6">
                {({ viewable }) => (
                    <div>
                        <a href="/" aria-hidden={viewable ? null : true} tabIndex={viewable ? null : "-1"}>
                            <Image src="/computer-user.jpeg" alt="A person sat down using a laptop computer" />
                        </a>
                    </div>
                )}
            </Carousel.Item>
            <Carousel.Item label="Slide 6 of 6">
                <Image src="/canyon.jpeg" alt="Landscape view of a canyon" />
            </Carousel.Item>
        </Carousel>
    </Story>
</Canvas>
```

JackHowa replied:
gotcha, can add to storybook

JackHowa replied:
updated @visionbegin

brentmiller1973 commented on src/components/carousel/index.jsx:
```diff
@@ -182,9 +218,34 @@ const Carousel = ({
+	) : (
+		<DefaultStopAutoplayButton id={id} onClick={toggleAutoplay} />
+	);
+
```
@JackHowa I don't know with this is scope creep (I think it may be), but I would think it would be prudent that after the last slide is reached, the button text stops staying "Stop Autoplay" and instead says "Start Autoplay".  If it's pressed it can either start back at slide one or go in reverse....up to product.

JackHowa replied:
cool asked jenae in the ticket https://arcpublishing.atlassian.net/browse/TMEDIA-835?focusedCommentId=736246

nschubach replied:
Considering this works, could we discuss this and make it another story?  As far as I'm concerned, this is the only thing holding this up.

JackHowa replied:
@visionbegin this has been updated per the ticket. At the end of the cycle (after the 4 second interval like engine theme sdk), the autoplay will stop and stay at the last slide

JackHowa replied:
https://github.com/WPMedia/arc-themes-components/pull/73/commits/4fced71e28ff79cc6f10df8d089dab372fe0df9f

JackHowa replied:
@visionbegin this has been updated. is it ok to resolve?

brentmiller1973 reviewed (dismissed):
Looks great Jack!  Left two comments....the one on additional stories I think is a must, but the comment on what to do when the last slide is reached might be outside the scope of this ticket and may require some discussions with design and product.

JackHowa committed:
docs: Add example for full-screen mode as well as autoplay and multiple slides showing

Per feedback https://github.com/WPMedia/arc-themes-components/pull/73#discussion_r853202326

nschubach commented on index.js:
```diff
@@ -34,12 +36,14 @@ export {
 	Icon,
 	Image,
 	Input,
+	isServerSide,
 	Link,
 	MediaItem,
 	Overline,
 	Paragraph,
 	Pill,
 	Separator,
 	Stack,
+	useInterval,
```
Should we namespace utility functions or leave them intermixed with the components?

JackHowa replied:
in the past, we've intermixed them

JackHowa replied:
I don't really have an issue with doing so. makes exports easier

nschubach replied:
if we do, it would have to be another story, going to resolve this conversation.

nschubach committed:
fix 'undefined' classname

matthewroach commented:
@JackHowa Looking through the pull requests for the updates to the carousel with the additional controls above - has there been a change in design or product decision to order the controls in this new way vs the current design of Full Screen and Autoplay on the left, and label and additional buttons on the right?

matthewroach commented:
@JackHowa Looking through the pull requests for the updates to the carousel with the additional controls above - has there been a change in design or product decision to order the controls in this new way vs the current design of Full Screen and Autoplay on the left, and label and additional buttons on the right?

JackHowa commented:
@matthewroach will work on it; thanks for the note

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-689-autoplay-gallery

JackHowa committed:
Merge branch 'TMEDIA-689-autoplay-gallery' of https://github.com/WPMedia/arc-themes-components into TMEDIA-689-autoplay-gallery

JackHowa committed:
fix: Reorder and contain the actions per design feedback

Per https://github.com/WPMedia/arc-themes-components/pull/73#issuecomment-1103794266

JackHowa commented:
> @JackHowa Looking through the pull requests for the updates to the carousel with the additional controls above - has there been a change in design or product decision to order the controls in this new way vs the current design of Full Screen and Autoplay on the left, and label and additional buttons on the right?

@matthewroach updated the order and their containers. 

<img width="737" alt="Screen Shot 2022-04-20 at 09 23 11" src="https://user-images.githubusercontent.com/5950956/164252869-03727f3d-f04c-46d0-959e-c74c1cb980fb.png">

-  [x] Will work on the feature change of only labels and icons for buttons rather than custom buttons and a fallback like next/previous
- [x] Oh I see. The autoplay does not immediately toggle off in the button when it hits the end. you have to wait until the next 4 seconds interval for that to happen. Will update autoplay stop functionality

JackHowa committed:
feat: End autoplay on end of the line of the slides per engine theme sdk

Per comments in ticket https://arcpublishing.atlassian.net/browse/TMEDIA-835?focusedCommentId=736676

JackHowa commented:
@matthewroach 

like this? 

```jsx
const AutoplayButton = ({ id, onClick, iconNode, text, ariaLabel }) => (
	<Button
		id={id}
		onClick={onClick}
		label={ariaLabel}
		className={`${COMPONENT_CLASS_NAME}__button ${COMPONENT_CLASS_NAME}__button--auto-play`}
	>
		{iconNode}
		{text}
	</Button>
);
```

JackHowa committed:
feat: Add aria label and icon for nodes and custom icons

JackHowa committed:
docs: Add docs for using own icon, etc per new requirements

JackHowa committed:
test: Adjsut the tests to accomodate for defaults change in aria label and text

JackHowa committed:
fix: Add back accidental removal. Fix chromatic

nschubach reviewed (approved):
👍

JackHowa committed:
feat: Add flex default so that the items are aligned next to each other horizontally

JackHowa committed:
feat: Center items vertically

JackHowa commented:
@matthewroach updated to align items within the controls container next to each other horizontally rather than vertically by default

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-689-autoplay-gallery

JackHowa committed:
fix: ENsure bad merge works with scss

JackHowa committed:
Merge branch 'arc-themes-release-version-2.00' into TMEDIA-689-autoplay-gallery

JackHowa committed:
fix: Ensure classname maintains the same style across states

JackHowa committed:
docs: Update the slide name for the label

JackHowa commented on src/components/carousel/index.jsx:
```diff
@@ -107,15 +126,17 @@ const Carousel = ({
 	slidesToShow,
-	fullScreenShowButton,
-	fullScreenMinimizeButton,
-	enableFullScreen,
```
updating alphabetical

ecarlisle reviewed (approved):
@JackHowa :  I think the code, testing, and Storybook examples are all great! The only question I had was when Start Autoplay was selected on the last slide. I see that's been brought up and you've asked Product about it here:
https://arcpublishing.atlassian.net/browse/TMEDIA-835?focusedCommentId=736246
Great job!

JackHowa commented:
@visionbegin saw you had ❤️ the post in slack about reviewing the pr. but will merge this for now as far as two approvals. I reached out with the comments and resolving those. I think they've been addressed so I will resolve them
